provider/datagen/src/transform/cldr/units/mod.rs

```diff
@@ -2,34 +2,136 @@
 };
-use icu_unitsconversion::provider::{UnitsConstantsV1, UnitsConstantsV1Marker};
-use zerovec::ZeroMap;
+use icu_unitsconversion::provider::{
+    ConstantExactness, ConstantValue, UnitsConstantsV1, UnitsConstantsV1Marker,
+};
+use zerovec::{ZeroMap, ZeroVec};
 
 impl DataProvider<UnitsConstantsV1Marker> for crate::DatagenProvider {
     fn load(&self, _req: DataRequest) -> Result<DataResponse<UnitsConstantsV1Marker>, DataError> {
         self.check_req::<UnitsConstantsV1Marker>(_req)?;
 
-        let _units_data: &cldr_serde::units::units_constants::Resource = self
+        let units_data: &cldr_serde::units::units_constants::Resource = self
             .cldr()?
             .core()
             .read_and_parse("supplemental/units.json")?;
-        let mut constants_map = BTreeMap::<&str, &str>::new();
+        let constants = &units_data.supplemental.unit_constants.constants;
+
+        let mut constants_map_in_str_form =
+            BTreeMap::<&str, (Vec<String>, Vec<String>, ConstantExactness)>::new();
+        for (cons_name, cons_value) in constants {
+            let value = remove_whitespace(&cons_value.value);
+            let (num, den) = convert_constant_to_num_denom_strings(&value)?;
 
-        let constants = &_units_data.supplemental.unit_constants.constants;
-        for (key, constant) in constants {
-            constants_map.insert(key, &constant.value);
+            let constant_type = match cons_value.status.as_deref() {
+                Some("approximate") => ConstantExactness::Approximate,
+                _ => ConstantExactness::Actual,
+            };
+
+            constants_map_in_str_form.insert(cons_name, (num, den, constant_type));
         }
 
-        let result = UnitsConstantsV1 {
-            constants_map: ZeroMap::from_iter(constants_map),
-        };
+        // This loop iterates over the constants, replacing any string values with their corresponding constant values.
+        let mut has_internal_constants;
+        loop {
+            has_internal_constants = false;
+            let mut constants_with_constants_map_replaceable =
+                BTreeMap::<&str, (Vec<String>, Vec<String>, ConstantExactness)>::new();
+            for (cons_name, (num, den, constant_type)) in constants_map_in_str_form.iter() {
+                let mut temp_num = num.clone();
+                let mut temp_den = den.clone();
+                let mut temp_constant_type = *constant_type;
+
+                for i in 0..temp_num.len() {
+                    if !contains_alphabetic_chars(temp_num[i].as_str())
+                        || is_scientific_number(temp_num[i].as_str())
+                    {
+                        continue;
+                    }
+
+                    has_internal_constants = true;
+                    if let Some((rnum, rden, rconstant_type)) =
+                        constants_map_in_str_form.get(temp_num[i].as_str())
+                    {
+                        temp_num.remove(i);
```

Comment:
You're iterating over `temp_num` by index but removing elements on the way, so I think you might be skipping elements:

* i = 0, temp_num = [foo, bar] (start)
* i = 0, temp_num = [bar, 1, 7] (replace foo)
* i = 1, temp_num = [bar, 1, 7] (next iteration, skips bar)
* done